File: SnowplowDemo/SnowplowDemo/ViewController.m
//
//  ViewController.m
//  SnowplowDemo
//
//  Copyright (c) 2015-2018 Snowplow Analytics Ltd. All rights reserved.
//
//  This program is licensed to you under the Apache License Version 2.0,
//  and you may not use this file except in compliance with the Apache License
//  Version 2.0. You may obtain a copy of the Apache License Version 2.0 at
//  http://www.apache.org/licenses/LICENSE-2.0.
//
//  Unless required by applicable law or agreed to in writing,
//  software distributed under the Apache License Version 2.0 is distributed on
//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
//  express or implied. See the Apache License Version 2.0 for the specific
//  language governing permissions and limitations there under.
//
//  Authors: Joshua Beemster
//  Copyright: Copyright (c) 2015-2018 Snowplow Analytics Ltd
//  License: Apache License Version 2.0
//

#import "ViewController.h"
#import "DemoUtils.h"
#import "SPTracker.h"
#import "SPEmitter.h"
#import "SPUtilities.h"
#import "SPSubject.h"

@interface ViewController ()

@property (nonatomic, weak) IBOutlet UILabel *            madeLabel;
@property (nonatomic, weak) IBOutlet UILabel *            dbCountLabel;
@property (nonatomic, weak) IBOutlet UILabel *            isRunningLabel;
@property (nonatomic, weak) IBOutlet UILabel *            isOnlineLabel;
@property (nonatomic, weak) IBOutlet UILabel *            sentCountLabel;
@property (nonatomic, weak) IBOutlet UILabel *            sessionCountLabel;
@property (nonatomic, weak) IBOutlet UILabel *            isBackgroundLabel;
@property (nonatomic, weak) IBOutlet UITextField *        urlTextField;
@property (nonatomic, weak) IBOutlet UISegmentedControl * methodType;
@property (nonatomic, weak) IBOutlet UISegmentedControl * trackingOnOff;
@property (nonatomic, weak) IBOutlet UISegmentedControl * protocolType;
@property (strong, nonatomic) IBOutlet UIScrollView *     scrollView;

@end

@implementation ViewController {
    SPTracker *       _tracker;
    long long int     _madeCounter;
    long long int     _sentCounter;
    NSTimer *         _updateTimer;
}

- (void) viewDidLoad {
    [super viewDidLoad];
    [self setup];
}

- (void) didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
}

- (void) setup {
    _tracker = [self getTrackerWithUrl:@"http://acme.fake.com" method:SPRequestPost];
    _updateTimer = [NSTimer scheduledTimerWithTimeInterval:1 target:self selector:@selector(updateMetrics) userInfo:nil repeats:YES];
    _urlTextField.delegate = self;
    [_trackingOnOff addTarget:self
                       action:@selector(action)
             forControlEvents:UIControlEventValueChanged];
}

- (IBAction) trackEvents:(id)sender {
    NSString *url = [self getCollectorUrl];
    SPRequestOptions methodType = [self getMethodType];
    SPProtocol protocolType = [self getProtocolType];
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        if ([url isEqual: @""]) {
            return;
        }
        
        // Update the Tracker
        
        // Ensures the application won't crash with a bad URL
        @try {
            [_tracker.emitter setUrlEndpoint:url];
        }
        @catch (NSException *exception) {
            return;
        }
        
        [_tracker.emitter setHttpMethod:methodType];
        [_tracker.emitter setProtocol:protocolType];
        
        // Itterate the made counter
        _madeCounter += 25;
        
        // Track all types of events!
        [DemoUtils trackAll:_tracker];
    });
}

- (void) updateMetrics {
    [_madeLabel setText:[NSString stringWithFormat:@"Made: %lld", _madeCounter]];
    [_dbCountLabel setText:[NSString stringWithFormat:@"DB Count: %lu", (unsigned long)[_tracker.emitter getDbCount]]];
    [_sessionCountLabel setText:[NSString stringWithFormat:@"Session Count: %lu", (unsigned long)[_tracker getSessionIndex]]];
    [_isRunningLabel setText:[NSString stringWithFormat:@"Running: %s", [_tracker.emitter getSendingStatus] ? "yes" : "no"]];
    [_isBackgroundLabel setText:[NSString stringWithFormat:@"Background: %s", [_tracker getInBackground] ? "yes" : "no"]];
    [_isOnlineLabel setText:[NSString stringWithFormat:@"Online: %s", [SPUtilities isOnline] ? "yes" : "no"]];
    [_sentCountLabel setText:[NSString stringWithFormat:@"Sent: %lu", (unsigned long)_sentCounter]];
}

- (void) action {
    BOOL tracking = _trackingOnOff.selectedSegmentIndex == 0 ? YES : NO;
    if (tracking && ![_tracker getIsTracking]) {
        [_tracker resumeEventTracking];
    } else if ([_tracker getIsTracking]) {
        [_tracker pauseEventTracking];
    }
}

- (NSString *) getCollectorUrl {
    return _urlTextField.text;
}

- (enum SPRequestOptions) getMethodType {
    return _methodType.selectedSegmentIndex == 0 ? SPRequestGet : SPRequestPost;
}

- (enum SPProtocol) getProtocolType {
    return _protocolType.selectedSegmentIndex == 0 ? SPHttp : SPHttps;
}

- (BOOL) textFieldShouldReturn:(UITextField *)textField {
    return [textField resignFirstResponder];
}

static NSString *const kAppId     = @"DemoAppId";
static NSString *const kNamespace = @"DemoAppNamespace";

// Tracker Setup & Init

- (SPTracker *) getTrackerWithUrl:(NSString *)url_
                           method:(enum SPRequestOptions)method_ {
    SPEmitter *emitter = [SPEmitter build:^(id<SPEmitterBuilder> builder) {
        [builder setUrlEndpoint:url_];
        [builder setHttpMethod:method_];
        [builder setCallback:self];
        [builder setEmitRange:500];
        [builder setEmitThreadPoolSize:20];
        [builder setByteLimitPost:52000];
    }];
    
    SPSubject *subject = [[SPSubject alloc] initWithPlatformContext:YES andGeoContext:NO];
    
    SPTracker *tracker = [SPTracker build:^(id<SPTrackerBuilder> builder) {
        [builder setEmitter:emitter];
        [builder setAppId:kAppId];
        [builder setTrackerNamespace:kNamespace];
        [builder setBase64Encoded:false];
        [builder setSessionContext:YES];
        [builder setSubject:subject];
    }];
    return tracker;
}

// Define Callback Functions

- (void) onSuccessWithCount:(NSInteger)successCount {
    _sentCounter += successCount;
}

- (void) onFailureWithCount:(NSInteger)failureCount successCount:(NSInteger)successCount {
    _sentCounter += successCount;
}

@end

File: SnowplowSwiftDemo/SnowplowSwiftDemo/ViewController.swift
//
//  ViewController.swift
//  SnowplowSwiftDemo
//
//  Created by Michael Hadam on 1/17/18.
//  Copyright © 2018 snowplowanalytics. All rights reserved.
//

import UIKit
import Foundation
import CoreData
import SwiftTryCatch
import SnowplowTracker

class ViewController: UIViewController, UITextFieldDelegate, SPRequestCallback {
    @IBOutlet weak var uriField: UITextField!
    @IBOutlet weak var trackingSwitch: UISegmentedControl!
    @IBOutlet weak var protocolSwitch: UISegmentedControl!
    @IBOutlet weak var methodSwitch: UISegmentedControl!
    @IBOutlet weak var isRunningLabel: UILabel!
    @IBOutlet weak var isBackgroundLabel: UILabel!
    @IBOutlet weak var sessionCountLabel: UILabel!
    @IBOutlet weak var isOnlineLabel: UILabel!
    @IBOutlet weak var madeLabel: UILabel!
    @IBOutlet weak var dbCountLabel: UILabel!
    @IBOutlet weak var sentLabel: UILabel!
    
    var tracker : SPTracker!
    var madeCounter : Int = 0
    var sentCounter : Int = 0
    var updateTimer : Timer!
    var uri : String = ""
    var methodType : SPRequestOptions = SPRequestGet
    var protocolType : SPProtocol = SPHttp
    
    let kAppId     = "DemoAppId"
    let kNamespace = "DemoAppNamespace"

    
    func setup() {
        self.tracker = self.getTracker("acme.fake.com", method: SPRequestPost)
        self.updateTimer = Timer.scheduledTimer(timeInterval: 1, target: self, selector: #selector(updateMetrics), userInfo: nil, repeats: true)
        self.uriField.delegate = self
        self.trackingSwitch.addTarget(self, action: #selector(action), for: .valueChanged)
    }
    
    @objc func action() {
        let tracking: Bool = (trackingSwitch.selectedSegmentIndex == 0)
        if (tracking && !(tracker.getIsTracking())) {
            tracker.resumeEventTracking()
        } else if (tracker.getIsTracking()) {
            tracker.pauseEventTracking()
        }
    }
    
    func getCollectorUrl() -> String {
        return self.uri
    }
    
    func getMethodType() -> SPRequestOptions {
        return self.methodType
    }
    
    func getProtocolType() -> SPProtocol {
        return self.protocolType
    }
    
    func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        self.uri = uriField.text!
        return textField.resignFirstResponder()
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        self.setup()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @IBAction func inputUri(_ sender: UITextField) {
        self.uri = uriField.text!
    }
    
    @IBAction func toggleMethod(_ sender: UISegmentedControl) {
        self.methodType = (methodSwitch.selectedSegmentIndex == 0) ?
            SPRequestGet : SPRequestPost
    }
    
    @IBAction func toggleProtocol(_ sender: UISegmentedControl) {
        self.protocolType = (protocolSwitch.selectedSegmentIndex == 0) ?
            SPHttp : SPHttps
    }
    
    @IBAction func trackEvents(_ sender: UIButton) {
        DispatchQueue.global(qos: .default).async {
            let url = self.getCollectorUrl()
            if url == "" {
                return
            }
            
            // Update the tracker
            
            SwiftTryCatch.try({
                self.tracker.emitter.setUrlEndpoint(url)
            }, catch: { (error) in
                return
            }, finally: {
                
            })
            
            self.tracker.emitter.setHttpMethod(self.getMethodType())
            self.tracker.emitter.setProtocol(self.getProtocolType())
            
            // Iterate the made counter
            self.madeCounter += 28;
            
            // Track all types of events
            DemoUtils.trackAll(self.tracker)
        }
    }
    
    @objc func updateMetrics() {
        madeLabel.text = String(format: "Made: %lld", madeCounter)
        dbCountLabel.text = String(format: "DB Count: %lu", CUnsignedLong(self.tracker.emitter.getDbCount()))
        sessionCountLabel.text = String(format: "Session Count: %lu", CUnsignedLong(self.tracker.getSessionIndex()))
        isRunningLabel.text = String(format: "Running: %@", self.tracker.emitter.getSendingStatus() ? "yes" : "no")
        isBackgroundLabel.text = String(format: "Background: %@", self.tracker.getInBackground() ? "yes" : "no")
        isOnlineLabel.text = String(format: "Online: %@", SPUtilities.isOnline() ? "yes" : "no")
        sentLabel.text = String(format: "Sent: %lu", CUnsignedLong(sentCounter))
    }
    
    // Tracker setup and init
    
    func getTracker(_ url: String, method: SPRequestOptions) -> SPTracker {
        let emitter = SPEmitter.build({ (builder : SPEmitterBuilder?) -> Void in
            builder!.setUrlEndpoint(url)
            builder!.setHttpMethod(method)
            builder!.setCallback(self)
            builder!.setEmitRange(500)
            builder!.setEmitThreadPoolSize(20)
            builder!.setByteLimitPost(52000)
        })
        let subject = SPSubject(platformContext: true, andGeoContext: false)
        let newTracker = SPTracker.build({ (builder : SPTrackerBuilder?) -> Void in
            builder!.setEmitter(emitter)
            builder!.setAppId(self.kAppId)
            builder!.setTrackerNamespace(self.kNamespace)
            builder!.setBase64Encoded(false)
            builder!.setSessionContext(true)
            builder!.setSubject(subject)
        })
        return newTracker!
    }
    
    // Define Callback Functions
    
    func onSuccess(withCount successCount: Int) {
        self.sentCounter += successCount;
    }
    
    func onFailure(withCount failureCount: Int, successCount: Int) {
        self.sentCounter += successCount;
    }
}
